persistent-postgresql/Database/Persist/Postgresql.hs

```diff
@@ -1712,16 +1712,70 @@ upsertWhere
 upsertWhere record updates filts =
   upsertManyWhere [record] [] updates filts
 
--- | Postgres specific 'upsertManyWhere'. This method does the following:
--- It will insert a record if no matching unique key exists.
--- If a unique key exists, it will update the relevant field with a user-supplied value, however,
--- it will only do this update on a user-supplied condition.
--- For example, here's how this method could be called like such:
+-- | Postgres specific 'upsertManyWhere'.
 --
--- upsertManyWhere [record] [recordField =. newValue] [recordField !=. newValue]
+-- The first argument determines a list of new records to insert.
 --
--- Called thusly, this method will insert a new record (if none exists) OR update a recordField with a new value
--- assuming the condition in the last block is met.
+-- If there's unique key collisions for some or all of the proposed insertions,
+-- you can use the second argument to specify which fields, under which
+-- conditions (using 'HandleUpdateCollision' strategies,) will be copied from
+-- each proposed, conflicting new record onto its corresponding row already present
+-- in the database. Helpers such as 'copyField', 'copyUnlessEq',
+-- 'copyUnlessNull' and 'copyUnlessEmpty' are exported from this module to help
+-- with this, as well as the constructors for 'HandleUpdateCollision'.
+-- For example, you may have a patch of updates to apply, some of which are NULL
+-- to represent "no change"; you'd use 'copyUnlessNull' for each applicable
+-- field in this case so any new values in the patch which are NULL don't
+-- accidentally overwrite existing values in the database which are non-NULL.
```

Comment:
This wording is a little confusing to me (but, hey, the function itself is a bit confusing!) I think if I were to write it, I'd say something like:

> The second list describes what happens if one of the records to be inserted conflicts with a unique key in the database. The 'HandleUpdateCollision' type has more documentation on this. Briefly, we can:
> 
> 1. `copyField` will copy the value from the record we tried to insert into the record in the database table.
> 2. `copyUnlessEq` (describe this)
> 3. `copyUnlessNull` (describe this)
> 4. `etc`

Reply:
I like the more abridged approach — I'm not 100% sure but, if I follow [the docs](https://www.postgresql.org/docs/current/sql-insert.html#SQL-ON-CONFLICT) I think the `one of the records` part might be a lil bit misleading, since there could be multiple collisions each of which will be resolved in the `ON CONFLICT` clause (in my mind `one of` means that only one collision would be resolved? Maybe I'm reading too deep into it.) 

Maybe something like `what happens for a record proposed for insertion whose unique key conflicts with an existing record in the database` (as you can see, the more precise I attempt to get, the more I veer into verbosity lol)

Reply:
That's a great point! Thanks. 

It is a rather tricky thing to articulate!

> The second list describes what happens to the records which have unique key conflicts with existing rows in the database. The `HandleUpdateCollision` type documents this more thoroughly. Briefly,
>
> 1. `copyField` will update the row in the database with the value of the field of the record we attempted to insert.

is another attempt, but I'm not thrilled with it.

But let me try and break this down:

1. We have a `[record]` that we want to `upsert` into the database.
2. If the `record` does not exist, then we `insert` it.
3. If the `record` already exists by some `Unique record`, then we use `[HandleUpdateCollision record]` to specify how to handle the collision for that record.

You know, it's occuring to me that this is *much* easier to express in the singular. For example, `upsertWhere :: rec -> [HandleUpdateCollision rec] -> [Update rec] -> [Filter rec] -> m ()`. We can talk about that like:

> Attempt to insert the `record` into the database. If there is a unique key conflict, and the `[Filter rec]` are satisfied for the record we are attempting to insert, then we use the `[HandleUpdateCollision record]` and `[Update rec]` to specify how to update the record.

I think that `[Update rec]` aren't included in `upsertWhere` because it's easy enough to just edit the record itself and copy the field in.

But if `upsertWhere` is documented in an easy to explain manner, then we can make `upsertManyWhere` as:

> Like `upsertWhere`, but does a batch upsert for all of the records in the list.